packages/lodestar/src/chain/opPools/aggregatedAttestationPool.ts

```diff
@@ -0,0 +1,216 @@
+  // no need to maintain lowestPermissibleSlot since it's checked via gossip validation fn
+  add(attestation: phase0.Attestation, attestingIndices: ValidatorIndex[], committee: ValidatorIndex[]): InsertOutcome {
+    const slot = attestation.data.slot;
+    const attestationGroupByDataHash = this.attestationGroupByDataHashBySlot.getOrDefault(slot);
+    const dataRoot = ssz.phase0.AttestationData.hashTreeRoot(attestation.data);
+    const dataRootHex = toHexString(dataRoot);
+    let attestationGroup = attestationGroupByDataHash.get(dataRootHex);
+    if (attestationGroup) {
+      return attestationGroup.add({attestation, attestingIndices});
+    } else {
+      attestationGroup = new MatchingDataAttestationGroup({attestation, attestingIndices}, committee);
+      attestationGroupByDataHash.set(dataRootHex, attestationGroup);
+      return InsertOutcome.NewData;
+    }
+  }
+
+  /** Remove attestations included in the chain, return number of attestations removed */
+  removeIncluded(attestation: phase0.Attestation, attestingIndices: ValidatorIndex[]): number {
+    const slot = attestation.data.slot;
+    const attestationGroupByDataHash = this.attestationGroupByDataHashBySlot.get(slot);
+    if (!attestationGroupByDataHash) {
+      return 0;
+    }
+    const dataRoot = ssz.phase0.AttestationData.hashTreeRoot(attestation.data);
+    const dataRootHex = toHexString(dataRoot);
+    const attestationGroup = attestationGroupByDataHash.get(dataRootHex);
+    if (!attestationGroup) {
+      return 0;
+    }
+    return attestationGroup.removeIncluded({attestation, attestingIndices});
+  }
+
+  /** Remove attestations which are too old to be included in a block. */
+  prune(clockSlot: Slot): void {
+    // Only retain SLOTS_PER_EPOCH slots
+    pruneBySlot(this.attestationGroupByDataHashBySlot, clockSlot, SLOTS_PER_EPOCH);
+  }
+
+  /**
+   * Get attestations to be included in a block.
+   * Attestations are sorted by inclusion distance then number of attesters.
+   * Attestations should pass the validation when processing attestations in beacon-state-transition.
+   */
+  getAttestationsForBlock(state: CachedBeaconState<allForks.BeaconState>): phase0.Attestation[] {
+    const slots = Array.from(this.attestationGroupByDataHashBySlot.keys()).sort((a, b) => b - a);
+    const attestations: phase0.Attestation[] = [];
+    for (const slot of slots) {
+      const attestationGroupByDataHash = this.attestationGroupByDataHashBySlot.get(slot);
+      // should not happen
+      if (!attestationGroupByDataHash) {
+        throw Error(`No aggregated attestation pool for slot=${slot}`);
+      }
+      const attestationGroups = Array.from(attestationGroupByDataHash.values());
+      attestations.push(...attestationGroups.map((group) => group.getAttestations()).flat());
+    }
+    // consumer should limit MAX_ATTESTATIONS items
+    return attestations.filter((attestation) => safeValidateAttestation(state, attestation));
```

Comment:
Is this safe performance wise? Why do we need to re-validate the attestation here?

Reply:
this is part of the spec `The attestations added must satisfy the verification conditions found in attestation processing`, Teku and lighthouse both do that.

The `validateAttestation` is part of our state transition and contains some simple checks there https://github.com/ChainSafe/lodestar/blob/70949d5d74956c692b243578d9845b8ab0cc0eea/packages/beacon-state-transition/src/phase0/block/processAttestation.ts#L60

Reply:
I would ask to profile very well all this code in a testnet node or with performance tests to ensure we are good performance-wise

Reply:
It still feels like lots of repeated work. The attestation was validated before to get here, what conditions can be invalidated from then and now?

Reply:
one condition is `attestation.aggregationBits.length !== committee.length`, another is `!(data.target.epoch === epochCtx.previousShuffling.epoch || data.target.epoch === epochCtx.currentShuffling.epoch)`

these were not checked in `validateGossipAggregateAndProof`